File: java/src/main/java/io/github/ivere27/synurang/PluginStream.java
package io.github.ivere27.synurang;

/**
 * A stream handle for streaming RPCs with a Synurang plugin.
 * <p>
 * Supports send, recv, closeSend, and close operations.
 * The recv() method returns null on EOF.
 * <p>
 * Response format: [status:1byte][payload...]
 * status=0: success, payload is protobuf
 * status=1: error, payload is error message
 */
public class PluginStream implements AutoCloseable {
    private final PluginHost host;
    private final long handle;
    private final PluginHost.StreamFuncs funcs;
    private final java.util.concurrent.atomic.AtomicBoolean closed = new java.util.concurrent.atomic.AtomicBoolean(false);

    PluginStream(PluginHost host, long handle, PluginHost.StreamFuncs funcs) {
        this.host = host;
        this.handle = handle;
        this.funcs = funcs;
    }

    /**
     * Send data to the stream.
     *
     * @param data serialized protobuf message
     * @throws PluginError on send failure
     */
    public void send(byte[] data) throws PluginError {
        if (closed.get()) throw new PluginError.ClosedError();

        int result = SynurangJni.nativeStreamSend(funcs.send, handle, data);
        if (result != 0) {
            throw new PluginError("Stream send failed with code " + result);
        }
    }

    /**
     * Receive data from the stream.
     * <p>
     * The status byte is checked and stripped. Returns the protobuf payload.
     *
     * @return protobuf response bytes (status byte stripped), or null on EOF
     * @throws PluginError on error
     */
    public byte[] recv() throws PluginError {
        if (closed.get()) throw new PluginError.ClosedError();

        byte[] result = SynurangJni.nativeStreamRecv(funcs.recv, host.getFreePtr(), handle);

        // null means EOF
        if (result == null) {
            return null;
        }

        if (result.length == 0) {
            throw new PluginError("Empty stream response");
        }

        // Check status byte
        if (result[0] == 1) {
            throw new PluginError(new String(result, 1, result.length - 1));
        }

        // Strip status byte
        byte[] payload = new byte[result.length - 1];
        System.arraycopy(result, 1, payload, 0, payload.length);
        return payload;
    }

    /**
     * Close the send side of the stream.
     * The stream can still receive data after this.
     */
    public void closeSend() {
        if (!closed.get()) {
            SynurangJni.nativeStreamCloseSend(funcs.closeSend, handle);
        }
    }

    /**
     * Close the stream completely. Thread-safe via CAS.
     */
    @Override
    public void close() {
        if (closed.compareAndSet(false, true)) {
            SynurangJni.nativeStreamClose(funcs.close, handle);
        }
    }
}

File: csharp/Synurang/PluginStream.cs
using System;
using System.Runtime.InteropServices;
using System.Threading;

namespace Synurang;

/// <summary>
/// A stream handle for streaming RPCs with a Synurang plugin.
/// Supports Send, Recv, CloseSend, and Dispose operations.
/// Recv() returns null on EOF.
/// </summary>
public class PluginStream : IDisposable
{
    private readonly PluginHost _host;
    private readonly ulong _handle;
    private readonly PluginHost.StreamFuncs _funcs;
    private int _disposed;

    internal PluginStream(PluginHost host, ulong handle, PluginHost.StreamFuncs funcs)
    {
        _host = host;
        _handle = handle;
        _funcs = funcs;
    }

    /// <summary>
    /// Send data to the stream.
    /// </summary>
    public void Send(byte[] data)
    {
        ThrowIfClosed();
        if (!_host.TryEnterNativeCall(allowDuringDisposal: false))
            throw new PluginClosedError();

        IntPtr dataPtr = data.Length > 0 ? Marshal.AllocCoTaskMem(data.Length) : IntPtr.Zero;
        try
        {
            if (data.Length > 0)
                Marshal.Copy(data, 0, dataPtr, data.Length);
            int result = _funcs.Send(_handle, dataPtr, data.Length);
            if (result != 0)
                throw new FfiError("Stream send failed with code " + result);
        }
        finally
        {
            if (dataPtr != IntPtr.Zero)
                Marshal.FreeCoTaskMem(dataPtr);
            _host.ExitNativeCall();
        }
    }

    /// <summary>
    /// Receive data from the stream.
    /// Returns raw protobuf response bytes, or null on EOF.
    /// </summary>
    public byte[]? Recv()
    {
        ThrowIfClosed();
        if (!_host.TryEnterNativeCall(allowDuringDisposal: false))
            throw new PluginClosedError();

        try
        {
            IntPtr resultPtr = _funcs.Recv(_handle, out int respLen, out int status);

            if (status == 1)
            {
                _host.FreeNative(resultPtr);
                return null;
            }

            if (status < 0)
            {
                if (resultPtr != IntPtr.Zero && respLen > 0)
                {
                    try
                    {
                        byte[] errBytes = new byte[respLen];
                        Marshal.Copy(resultPtr, errBytes, 0, respLen);
                        throw FfiError.FromPayload(errBytes);
                    }
                    finally
                    {
                        _host.FreeNative(resultPtr);
                    }
                }

                _host.FreeNative(resultPtr);
                throw new FfiError("Stream recv failed with status " + status);
            }

            if (status != 0)
            {
                _host.FreeNative(resultPtr);
                throw new FfiError("Stream recv failed with status " + status);
            }

            try
            {
                if (resultPtr == IntPtr.Zero)
                {
                    if (respLen == 0)
                        return Array.Empty<byte>();
                    throw new FfiError("Plugin returned null for stream recv");
                }

                byte[] payload = new byte[respLen];
                if (payload.Length > 0)
                    Marshal.Copy(resultPtr, payload, 0, payload.Length);
                return payload;
            }
            finally
            {
                _host.FreeNative(resultPtr);
            }
        }
        finally
        {
            _host.ExitNativeCall();
        }
    }

    /// <summary>
    /// Close the send side of the stream.
    /// The stream can still receive data after this.
    /// </summary>
    public void CloseSend()
    {
        ThrowIfClosed();
        if (!_host.TryEnterNativeCall(allowDuringDisposal: false))
            throw new PluginClosedError();

        try
        {
            _funcs.CloseSend(_handle);
        }
        finally
        {
            _host.ExitNativeCall();
        }
    }

    public void Dispose()
    {
        CloseInternal(allowDuringDisposal: true);
    }

    internal void DisposeFromHost()
    {
        CloseInternal(allowDuringDisposal: true);
    }

    private void CloseInternal(bool allowDuringDisposal)
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0)
            return;

        try
        {
            if (_host.TryEnterNativeCall(allowDuringDisposal))
            {
                try
                {
                    _funcs.Close(_handle);
                }
                finally
                {
                    _host.ExitNativeCall();
                }
            }
        }
        finally
        {
            _host.UnregisterStream(_handle);
        }
    }

    private void ThrowIfClosed()
    {
        if (NativeLoader.VolatileRead(ref _disposed) != 0)
            throw new PluginClosedError();
    }
}
